launchflow/builds/ecr_builder.py

```diff
@@ -0,0 +1,250 @@
+import asyncio
+import base64
+from typing import IO, List, Literal, Optional
+
+from docker.client import from_env
+from docker.errors import BuildError  # type: ignore
+
+from docker import errors
+from launchflow import exceptions
+from launchflow.models.flow_state import AWSEnvironmentConfig
+from launchflow.workflows.utils import tar_source_in_memory
+
+
+def _write_build_logs(f: IO, log_stream):
+    for chunk in log_stream:
+        if "stream" in chunk:
+            f.write(chunk["stream"])
+
+
+def _get_build_status(client, build_id):
+    response = client.batch_get_builds(ids=[build_id])
+    if not response["builds"]:
+        raise ValueError("No build found for the provided build ID.")
+    build_status = response["builds"][0]["buildStatus"]
+    return build_status
+
+
+async def _poll_build_completion(client, build_id, poll_interval=10):
+    """
+    Polls the status of a build until it is completed or fails.
+
+    :param client: Boto3 CodeBuild client
+    :param build_id: ID of the build to poll
+    :param poll_interval: Time in seconds between each poll
+    """
+    while True:
+        build_status = _get_build_status(client, build_id)
+        if build_status in ["SUCCEEDED"]:
+            break
+        elif build_status in ["FAILED", "FAULT", "TIMED_OUT", "STOPPED"]:
+            raise ValueError(f"Build failed with status: {build_status}")
+        else:
+            await asyncio.sleep(poll_interval)  # Use asyncio.sleep for async waiting
+
+
+class ECRDockerBuilder:
+    def __init__(
+        self,
+        build_directory: str,
+        build_ignore: List[str],
+        build_log_file: IO,
+        ecr_repository: str,
+        launchflow_project_name: str,
+        launchflow_environment_name: str,
+        launchflow_service_name: str,
+        launchflow_deployment_id: str,
+        aws_environment_config: AWSEnvironmentConfig,
+    ):
+        self.build_directory = build_directory
+        self.build_ignore = build_ignore
+        self.build_log_file = build_log_file
+        self.ecr_repository = ecr_repository
+        self.launchflow_project_name = launchflow_project_name
+        self.launchflow_environment_name = launchflow_environment_name
+        self.launchflow_service_name = launchflow_service_name
+        self.launchflow_deployment_id = launchflow_deployment_id
+        self.aws_environment_config = aws_environment_config
+        self.docker_client = from_env()
```

Comment:
IIRC we wrap the other calls to from_env in a try / catch so we can raise a special Docker not installed error

Reply:
docker is in the pyproject.toml so should be fine: https://github.com/launchflow/launchflow/blob/a8cbd3a4989146527d13c9d11a5fab4ba85f140e/pyproject.toml#L26